robot/control/modules/motion-control/RobotModel.cpp

```diff
@@ -0,0 +1,134 @@
+#include "RobotModel.hpp"
+#include <cmath>
+
+namespace motion {
+
+Matrix<3> make_rotation_matrix(double phi) {
+    return (Matrix<3>() <<
+            std::cos(phi), -std::sin(phi), 0,
+            std::sin(phi), std::cos(phi), 0,
+            0, 0, 1
+        ).finished();
+}
+
+constexpr double radians(double degrees) {
+    return degrees * M_PI / 180;
+}
+
+RobotModel::RobotModel() {
+    // Geometry matrix
+    G = 1 / kWheelRadius * (Matrix<4, 3>() <<
+            -std::sin(kWheelTheta[0]), std::cos(kWheelTheta[0]), kRobotRadius,
+            -std::sin(kWheelTheta[1]), std::cos(kWheelTheta[1]), kRobotRadius,
+            -std::sin(kWheelTheta[2]), std::cos(kWheelTheta[2]), kRobotRadius,
+            -std::sin(kWheelTheta[3]), std::cos(kWheelTheta[3]), kRobotRadius
+        ).finished();
+
+    M = Vector<3>(kRobotMass, kRobotMass, kRobotInertia).asDiagonal();
+
+    // Calculate the pseudo-inverse of G
+    Eigen::CompleteOrthogonalDecomposition<Eigen::MatrixXd> GT_cod(G.transpose());
+    GT_inv = GT_cod.pseudoInverse();
+
+    // Calculate the effective wheel inertia, at the output shaft.
+    double wheel_inertia = kWheelInertia + kRotorInertia / kGearRatio / kGearRatio;
+
+    // Calculate the total effective mass matrix of the robot, including
+    // body mass/inertia and additional wheel/motor inertia.
+    Z = M + G.transpose() * wheel_inertia * G;
+    Z_inv = Z.inverse();
+
+    wheel_friction_mat = G.transpose() * Matrix<4, 4>::Identity() * G;
+}
+
+RobotTwist RobotModel::forward_dynamics(RobotTwist v, MotorVoltage u) {
+    // Calculate motor torques
+    Vector<4> motor_velocity = G * v / kGearRatio;
+    Vector<4> motor_torque =
+        kMotorTorqueConstant * (u - kMotorSpeedConstant * motor_velocity) / kMotorResistance;
+
+    Vector<4> wheel_torque = motor_torque / kGearRatio;
+    Vector<3> effort = G.transpose() * wheel_torque;
+
+    Matrix<3, 3> coriolis = (Matrix<3, 3>() <<
+        0, -v(2), 0,
+        v(2), 0, 0,
+        0, 0, 0
+        ).finished();
+
+    Matrix<3, 3> V = wheel_friction_mat + coriolis * M;
```

Comment:
Is this actually the coriolis force cross product?

Reply:
I think so. The `coriolis` matrix is the skew-symmetric representation of `omega x` (multiplication by `coriolis` is the same as the cross product with the angular velocity vector).

Reply:
I guess what's throwing me off is that the usual place you see the coriolis effect is when you are find the acceleration of a point moving relative to frame B, which is moving relative to frame A. I don't think that's what we're doing here with the frame definitions?

Reply:
Honestly I'm not sure I got the terminology right, but I think it applies? You've got the spinning reference frame of the robot and the inertial world frame, right?

Reply:
FYI, what this actually does is captures the behavior of the robot when it has both linear and angular velocity (it experiences force perpendicular to its linear velocity and curves off to the side).

Reply:
I believe the only extra thing is that the object has to have a non-zero velocity in both reference frames. In the non-inertial frame, the object will appear to have a secondary acceleration acting upon it. If you kick a ball and view it from the robot frame (with a non-zero angular velocity), you'll see the ball veer off to the left or right even though it has no other forces acting on it in the inertial frame.

If it has 0 velocity in one of the frames, it's probably a euler or centrifugal. Based on your edit, centrifugal might be a better term. It's like that sideways outwards force you feel in a car when you are turning, centripetal force is the one inwards

Reply:
Centrifugal is -omega cross (omega cross r)
Euler is alpha cross r
Centripetal is omega cross (omega cross r)
Coriolis is 2 omega cross v
All have a extra inertia coefficient to convert accelerations to forces